core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java

```diff
@@ -619,11 +620,16 @@ protected Type visitDereferenceExpression(DereferenceExpression node, StackableA
             RowType rowType = (RowType) baseType;
             String fieldName = node.getField().getValue();
 
+            int index = -1;
             Type rowFieldType = null;
             for (RowType.Field rowField : rowType.getFields()) {
-                if (fieldName.equalsIgnoreCase(rowField.getName().orElse(null))) {
+                String rowFieldName = rowField.getName().orElse(null);
+                if (fieldName.equalsIgnoreCase(rowFieldName)) {
```

Comment:
What if we have an expression like this `CAST(ROW(1, 2, 3) AS ROW(a BIGINT, A DOUBLE, c BIGINT)).c`. In this case we won't fail right ? Is it okay to evaluate them based on the field being accessed ?

cc @martint , @kasiafi